internal/queue/queue_test.go

```diff
@@ -0,0 +1,190 @@
+package queue
+
+import (
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestReadWrite(t *testing.T) {
+	q := NewMessageQueue()
+
+	// Reading from an empty queue should return ErrQueueEmpty
+	if _, err := q.Read(); err != ErrQueueEmpty {
+		t.Fatal("expected to receive `ErrQueueEmpty` for reading from empty queue")
+	}
+
+	// Write 1 to the queue and read this later.
+	if err := q.Write(1); err != nil {
+		t.Fatal(err)
+	}
+
+	// Read the value. Value will be dequeued.
+	if msg, err := q.Read(); err != nil || msg != 1 {
+		t.Fatal(err)
+	}
+
+	// We just read a value, now try and read again and verify that we get ErrQueueEmpty again.
+	if _, err := q.Read(); err != ErrQueueEmpty {
+		t.Fatal(err)
+	}
+
+	// Close the queue and verify that we get an error on write.
+	q.Close()
+	if err := q.Write(1); err != ErrQueueClosed {
+		t.Fatal(err)
+	}
+}
+
+func TestReadOrWaitClose(t *testing.T) {
+	q := NewMessageQueue()
+
+	go func() {
+		q.Write(1)
+		q.Write(2)
+		q.Write(3)
+		time.Sleep(time.Second * 5)
+		q.Close()
+	}()
+
+	time.Sleep(time.Second * 2)
+
+	read := 0
+	for {
+		if _, err := q.ReadOrWait(); err != nil {
+			if err == ErrQueueClosed && read == 3 {
+				break
+			}
+			t.Fatal(err)
+		}
+		read++
+	}
+}
+
+func TestReadOrWait(t *testing.T) {
+	q := NewMessageQueue()
+
+	go func() {
+		q.Write(1)
+		q.Write(2)
+		q.Write(3)
+		time.Sleep(time.Second * 5)
+		q.Write(4)
+	}()
+
+	time.Sleep(time.Second * 1)
```

Comment:
Would the go func below not suffice to remove this? There's no need to wait since the `ReadOrWait` does it for us

Reply:
Want to sleep so that we're positive the first three values (or at the very least one) has been written to test both cases for ReadOrWait. Two cases I'm mentioning being that there is a value already and we just read it, and the other being that we wait for a value to arrive.